frame/offences/src/migration.rs

```diff
@@ -0,0 +1,67 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2018-2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::*;
+
+mod deprecated {
+    use crate::{Config, Perbill, OffenceDetails, SessionIndex};
+    use frame_support::{decl_module, decl_storage};
+    use sp_std::prelude::*;
+
+    /// Type of data stored as a deferred offence
+    pub type DeferredOffenceOf<T> = (
+        Vec<OffenceDetails<<T as frame_system::Config>::AccountId, <T as Config>::IdentificationTuple>>,
+        Vec<Perbill>,
+        SessionIndex,
+    );
+
+    decl_storage! {
+        trait Store for Module<T: Config> as Indices {
+            /// Deferred reports that have been rejected by the offence handler and need to be submitted
+            /// at a later time.
+            pub DeferredOffences get(fn deferred_offences): Vec<DeferredOffenceOf<T>>;
+        }
+    }
+    decl_module! {
+        pub struct Module<T: Config> for enum Call where origin: T::Origin { }
+    }
```

Comment:
Why do we need this? we just need to (deprecated) delcare some storage. 
```suggestion
```

Reply:
Good point, but in short the trait **Store** needs a Module, otherwise it doesn't compile.
```
error[E0412]: cannot find type `Module` in this scope
  --> frame/offences/src/migration.rs:33:25
   |
33 |         trait Store for Module<T: Config> as Indices {
   |                         ^^^^^^ not found in this scope
```

I implemented it based on the [documentation best practices](https://substrate.dev/docs/en/knowledgebase/runtime/upgrades#writing-a-migration) and following @apopiak [deprecated types example](https://github.com/hicommonwealth/substrate/blob/5f3933f5735a75d2d438341ec6842f269b886aaa/frame/indices/src/migration.rs#L5-L22)

Reply:
yeah the `decl_storage` relies on it

Reply:
Well that sucks :D you can just use something like this to define a storage item: 

https://github.com/paritytech/substrate/blob/e60597dff0aa7ffad623be2cc6edd94c7dc51edd/frame/staking/src/lib.rs#L1041-L1062

(don't need to use the macro, just the type alias)

Reply:
Yes I saw this method in the staking example, but I wasn't sure about which standard to use so I decided to follow the official documentation. Personally I also prefer the `generate_storage_types` approach because it reduces the code footprint and runtime binary size, the `decl_module ` generates a lot of unused code.

Reply:
So should I leave it as it or use the `generate_storage_types` instead? In case of using the `generate_storage_types!` I suggest to moving it to some common utility crate.

Reply:
Proposal for a generic storage macro:
``` rust
macro_rules! generate_storage_types {
    ($pallet_name:ident, $name:ident => Value<$value:ty>) => {
        paste::paste! {
            struct [<$name Instance>];
            impl frame_support::traits::StorageInstance for [<$name Instance>] {
                fn pallet_prefix() -> &'static str {
                    stringify!($pallet_name)
                }
                const STORAGE_PREFIX: &'static str = stringify!($name);
            }
            type $name = StorageValue<[<$name Instance>], $value, ValueQuery>;
        }
    }
}

generate_storage_types!(Offence, DeferredOffences => Value<Vec<DeferredOffenceOf<T>>>); 
```

Reply:
You can't use `generate_storage_types ` as it is yet, I wanted to make a PR and create a generalized version of it that can create proper alias for Value, Map and DoubleMap. 

Outside the scope of this PR. For now you should just create 1 `StorageInstance` and the type alias manually.

Reply:
(If you wait a few hours I'll probably make a PR for this :D)

Reply:
Cool! I possible could you share the PR link here? I would like to review it :)
Meanwhile I'll write the tests for `on_runtime_upgrade` hook.

Reply:
https://github.com/paritytech/substrate/pull/8456

Reply:
This is okay as-is, resolving.